README.md
# Tabla de contenidos
- [Tabla de contenidos](#tabla-de-contenidos)
- [Sistema de gestion para empresas funerarias "Papilio"](#sistema-de-gestion-para-empresas-funerarias-papilio)
- [Configuracion del Backend](#configuracion-del-backend)
  - [Configuracion de la base de datos](#configuracion-de-la-base-de-datos)
  - [Configuracion de seguridad de la API](#configuracion-de-seguridad-de-la-api)
  - [Configuracion inicial servidor de API](#configuracion-inicial-servidor-de-api)
  - [Configuracion de informacion del frontEnd para la API](#configuracion-de-informacion-del-frontend-para-la-api)
  - [Configuracion de cuenta de administrador](#configuracion-de-cuenta-de-administrador)
  - [Configuracion del puerto de escucha del backend](#configuracion-del-puerto-de-escucha-del-backend)
- [Configuracion de Frontend](#configuracion-de-frontend)
- [Puesta en marcha](#puesta-en-marcha)
  - [Backend](#backend)
  - [Frontend](#frontend)
    - [depuracion](#depuracion)
    - [produccion](#produccion)


# Sistema de gestion para empresas funerarias "Papilio"

Trabajo de fin de estudios para la Universidad de la Rioja, modalidad On-Line

# Configuracion del Backend

## Configuracion de la base de datos
En el directorio _API-papilio/app/config/_ se deben agregar las configuraciones de la base de datos

    module.exports = {
        HOST: "localhost",
        USER: "root",
        PASSWORD: "root",
        DB: "papilio",
        dialect: "mysql",
        pool: {
        max: 5,
        min: 0,
        acquire: 30000,
        idle: 10000
        }
    };

## Configuracion de seguridad de la API

En el directorio _API-papilio/app/config/_ se deben cambiar el secreto para mejorar la seguridad de JWT

    module.exports = {
        secret: "PAPILIO-0.01"
    };

## Configuracion inicial servidor de API

En el archivo index.js del directorio API-papilio:


    db.sequelize.sync().then(() => {
        console.log(chalk.red('Drop and Resync Db'));
        //initial(); //permite crear filas en la tabla seleccionada
    });

debe cambiarse a:

    db.sequelize.sync({force:true}).then(() => {
        console.log(chalk.red('Drop and Resync Db'));
        initial(); //permite crear filas en la tabla seleccionada
    });

Para la primera ejecucion, una vez creados los detalles iniciales se regresa a los valores iniciales y se vuelve a ejecutar el servidor. Una vez hecho esto, se puede dejar ejecutando con nodemon o pm2.

## Configuracion de informacion del frontEnd para la API

Estas configuraciones se utilizan para informar al usuario en el mensaje de correo electronico, que el perfil ha sido creado y puede visitarlo en ese directorio.

En el archivo _admin.controller.js_ en el directorio _API-papilio/app/controller/_ en las lineas 21-23 actualizar segun sus propias configuraciones del servidor

    var base_url = "localhost"
    var frontend_url = "localhost:3000"
    var backend_url = "localhost:8081"

En la linea 617 se puede cambiar el mensaje informativo que recibirá el usuario en formato html, tambien el asunto.

    var message = {
            from: resp1.dataValues.user, // Sender address
            to: resp2.dataValues.email,         // List of recipients
            subject: "Nueva memoria creada", // Subject line
            html: "<h3>Saludos</h3><p>Se ha creado una nueva memoria en el siguiente <a href='http://" + frontend_url + "/Memorial/" + crear.slug + "'>enlace</a></p>" // HTML body
        };

## Configuracion de cuenta de administrador

En el directorio _API-papilio/_ modificar el archivo index.js en la linea 88 
username y email son el mismo campo

    Usuarios.create({
        nombre_usuario: papilio,
        username: papilio,
        email: papilio,
        password: bcrypt.hashSync("admin", 8)
    }).then(user => {
        user.setRoles("1").then(() => {
        });
    }).catch(err => {
        res.status(200).send({ error: err.message });
    });

## Configuracion del puerto de escucha del backend

En el directorio _API-papilio/_ modificar el archivo index en la linea 112

    const PORT = process.env.PORT || 8081;
    app.listen(PORT, () => {
    console.log(`Server is running on port ${PORT}.`);
    });

# Configuracion de Frontend

En el directorio _papilio/lib/_ el archivo Vars.js debe contener la direccion del backend

    global.server="http://localhost:8081/";
    global.base_url="http://localhost:3000/";

# Puesta en marcha

## Backend
usar _nodemon_ o _pm2_ en el directorio API-papilio

## Frontend

### depuracion

usar npm run dev en el directorio papilio en caso de hacer pruebas

### produccion

Primero hay que compilar la pagina con run build y luego ejecutar run start

npm run build
npm run start





Changelog.md
**Tabla de Contenidos**

- [Changelog](#changelog)
  - [17/05/2021](#17052021)
  - [18/05/2021](#18052021)
  - [19/05/2021](#19052021)
  - [20/05/2021](#20052021)
  - [21/05/2021](#21052021)
  - [21/05/2021](#21052021-1)
  - [22/05/2021](#22052021)
  - [26/05/2021](#26052021)
  - [27/05/2021](#27052021)
  - [28/05/2021](#28052021)
  - [29/05/2021](#29052021)
  - [30/05/2021](#30052021)
    - [**Finalizada la fase 1** (vease **Fases del proyecto**)](#finalizada-la-fase-1-vease-fases-del-proyecto)
  - [31/05/2021](#31052021)
  - [1/06/2021](#1062021)
  - [2/06/2021](#2062021)
  - [3/06/2021](#3062021)
  - [4/06/2021](#4062021)
  - [5/06/2021](#5062021)
  - [6/06/2021](#6062021)
  - [7/06/2021](#7062021)
  - [8/06/2021](#8062021)
  - [9/06/2021](#9062021)
  - [10/06/2021](#10062021)
  - [11/06/2021](#11062021)
    - [**Finalizada la fase 2** (vease **Fases del proyecto**)](#finalizada-la-fase-2-vease-fases-del-proyecto)
  - [12/06/2021](#12062021)
  - [12/06/2021](#12062021-1)
  - [13/06/2021](#13062021)
  - [14/06/2021](#14062021)
  - [15/06/2021](#15062021)
    - [**Finalizada la fase 3** (vease **Fases del proyecto**)](#finalizada-la-fase-3-vease-fases-del-proyecto)
  - [16/06/2021](#16062021)
  - [17/06/2021](#17062021)
  - [18/06/2021](#18062021)
  - [19/06/2021](#19062021)
  - [20/06/2021](#20062021)
  - [21/06/2021](#21062021)
  - [22/06/2021](#22062021)
  - [23/06/2021](#23062021)
    - [**Finalizada la fase 4** (vease **Fases del proyecto**)](#finalizada-la-fase-4-vease-fases-del-proyecto)
  - [24/06/2021](#24062021)
  - [25/06/2021](#25062021)
  - [28/06/2021](#28062021)
  - [29/06/2021](#29062021)
  - [30/06/2021](#30062021)
  - [02/07/2021](#02072021)
  - [08/07/2021](#08072021)
  - [Pendiente:](#pendiente)
  - [Propuestas de mejora](#propuestas-de-mejora)

# Changelog
## 17/05/2021
- Creacion de la estructura de paginas a seguir
- Creacion del modelo jerarquia de usuario
- Creacion del protocolo de seguridad
  
## 18/05/2021
- Creacion del sistema de login del usuario
  - Actualizacion del protocolo de login del backend
  - Declaracion de uso del sistema jwt para protocolo de seguridad
  - Declaracion del uso de REST para comunicaciones entre BackEnd y FrontEnd

## 19/05/2021
- Correcciones menores al sistema de login del FrontEnd
- Correcciones menores al sistema de login del BackEnd
- Agregado el sistema de roles para el Backend

## 20/05/2021
- Correcciones menores al NavBar
- Se establece un sistema para determinar el estado de la sesion del usuario mediante props
- Establecido sistema de login y soporte heredado navbar

| Componente | Envia                | Recibe               |
| ---------- | -------------------- | -------------------- |
| Pagina*    | ServerloggedIn,token |                      |
| Layout     | ServerloggedIn,token | ServerloggedIn,token |
| NavBar     |                      | ServerloggedIn,token |



\* **Cualquier pagina que llame a los componentes hijos a continuacion**

- El menu de usuario cambia segun el tipo de usuario

**Usuario Visitante**
- Sobre Nosotros
- Ofertas
- Quienes somos
- Iniciar Sesion

**Usuario Normal**
- Ofertas
- Quienes Somos
- Perfil
- Cerrar Sesion

**Usuario Administrador**
- Perfil
- Ofertas
- Archivos
- Administrador
- Cerrar Sesion

## 21/05/2021
- Se crea el componente AdminBar, el cual es cargado unicamente por los componentes que lo necesiten, asi que no tiene variables de entrada
- Se modifica el AdminBar para que funcione con NextJs
- Se agregan los menues Sistema, Usuarios y ofertas
- La pagina de inicio "/" en caso de iniciar sesion como administrador, muestra el dashboard en el menu de configurar Sistema

## 21/05/2021
- Se agregan componentes a la pagina de inicio para cuando el usuario es administrador
- Bajo el menu Sistema/Configuracion se despliega el formulario que permite editar las configuraciones relacionadas con el SEO
- se agregan los campos relacionados con el backend:
  - **Nombre de la empresa:** Utilizado en el sistema SEO
  - **Nombres del propietario:** Utilizado en el formulario de contacto
  - **Apellidos del propietario:** Utilizadn en el formulario de contacto
  - **Descripcion del Negocio:** Utilizado en el sistema SEO y formulario de contacto
  - **Numero de contacto 1:** Utilizado en el sistema SEO y formulario de contacto
  - **Numero de contacto 2:** (Opcional) Utilizado en el sistema SEO y formulario de contacto
  - **Direccion del negocio:** Utilizado en el sistema SEO, Formulario de contacto y localizacion
  - **Correo Electronico 1:** Utilizado en el sistema SEO, Formulario de contacto y localizacion
  - **Correo Electronico 2:** (opcional)Utilizado en el sistema SEO, Formulario de contacto y localizacion
- Agregado submenu de seguridad al menu de Sistema
- Agregada tabla de configuracion general del Sistema

| Archivo                         | Tipo        | Funcion                                                               |
| ------------------------------- | ----------- | --------------------------------------------------------------------- |
| models/sistema.model.js         | modelo      | Tabla de configuracion del sistema                                    |
| routes/admin.routes.js          | rutas       | Ruta de llamadas de api para usuarios con rol de administracion       |
| controllers/admin.controller.js | controlador | Controlador para las llamadas de api mediante la ruta admin.routes.js |

## 22/05/2021
- Reconstruido sistema de comunicacion con la API

**Headers**
| nombre         | valor            |
| -------------- | ---------------- |
| Content-Type   | application/json |
| x-access-token | token            |

**Body**

**tipo:** raw

**valor**

    {
         "donde": "", //Lista de columnas en caso de busqueda, separadas por signo igual y caracter de tuberia (ex: id_usuario=1|nombre="juanito piguave")
         "limite": "", //funciona normal
         "off":"", //cuantos cosos por pagina 

API-papilio/index.js
const express = require("express");
const bodyParser = require("body-parser");
const cors = require("cors");
const chalk = require('chalk');
const app = express();
var bcrypt = require("bcryptjs");

const db = require("./app/models");
const SendmailTransport = require("nodemailer/lib/sendmail-transport");
const Roles = db.roles;
const Sistemas = db.sistemas;
const Sendmail = db.sendmail;
const Usuarios = db.usuarios;
// const tipos_tienda = db.tipos_tienda;
// const formas_de_pago = db.formas_de_pago;

var corsOptions = {
  origin: "http://localhost:8081"
};

//app.use(cors(corsOptions));
app.use(cors());

// parse requests of content-type - application/json
app.use(bodyParser.json());

// parse requests of content-type - application/x-www-form-urlencoded
app.use(bodyParser.urlencoded({ extended: true }));

//servir archivos estaticos
app.use("/img/", express.static('./uploads/'))
//establecemos las rutas 
require('./app/routes/auth.routes')(app);
require('./app/routes/admin.routes')(app);
require('./app/routes/public.routes')(app);
require('./app/routes/inicio.routes')(app);
require('./app/routes/user.routes')(app);
//require('./app/routes/mensajes.routes')(app);
// ruta de inicio
console.log(chalk.red("Inicializando sistema de BackEnd 'Papilio'"))
app.get("/", (req, res) => {
  res.json({ message: "No se supone que debas estar aquí" });
});
//sync({force:true}) permite dropear la base de datos entera, sin eso sincroniza la base con el sistema
db.sequelize.sync().then(() => {
  console.log(chalk.red('Drop and Resync Db'));

  //initial(); //permite crear filas en la tabla seleccionada
});

function initial() {
  Roles.create({
    id_rol: 1,
    nombre: "Admin"
  }).catch(function (err) {
    console.log(err);
  })

  Roles.create({
    id_rol: 2,
    nombre: "Usuario"
  }).catch(function (err) {
    console.log(err);
  })

  Roles.create({
    id_rol: 3,
    nombre: "Invitado"
  }).catch(function (err) {
    console.log(err);
  })

  Sistemas.create({
    nombre_empresa: "Papilio",
    nombres_propietario: "default",
    apellidos_propietario: "default",
    telefono1: "0000000000",
    correo1: "algo@algomas.com",
    direccion: "default",
    descripcion: "default",
  }).catch(function (err) {
    console.log(err);
  })

  Sendmail.create({
    host: "sendmail@papilio.com",
    port: "26",
    user: "sendmail@papilio.com",
    pass: "password",
    mensaje_registro: "<h3>Bienvenido a Papilio</h3>",
    mensaje_general: "<h3>Mensaje general</h3>",
  }).catch(function (err) {
    console.log(err);
  })

  Usuarios.create({
    nombre_usuario: "papilio",
    username: "papilio",
    email: "papilio",
    password: bcrypt.hashSync("admin", 8)
  }).then(user => {
    user.setRoles("1").then(() => {
    });
  }).catch(err => {
    res.status(200).send({ error: err.message });
  });


}

// set port, listen for requests
const PORT = process.env.PORT || 8081;
app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}.`);
});